src/main/java/com/scalar/db/server/DistributedTransactionService.java

```diff
@@ -0,0 +1,292 @@
+          responseObserver.onCompleted();
+        },
+        responseObserver);
+  }
+
+  @Override
+  public void commit(CommitRequest request, StreamObserver<Empty> responseObserver) {
+    execute(
+        () -> {
+          DistributedTransaction transaction = getTransaction(request.getTransactionId());
+          try {
+            transaction.commit();
+          } finally {
+            activeTransactions.remove(transaction.getId());
+          }
+          responseObserver.onNext(Empty.newBuilder().build());
+          responseObserver.onCompleted();
+        },
+        responseObserver);
+  }
+
+  @Override
+  public void abort(AbortRequest request, StreamObserver<Empty> responseObserver) {
+    execute(
+        () -> {
+          DistributedTransaction transaction = getTransaction(request.getTransactionId());
+          try {
+            transaction.abort();
+          } finally {
+            activeTransactions.remove(transaction.getId());
+          }
+          responseObserver.onNext(Empty.newBuilder().build());
+          responseObserver.onCompleted();
+        },
+        responseObserver);
+  }
+
+  @Override
+  public void getState(
+      GetTransactionStateRequest request,
+      StreamObserver<GetTransactionStateResponse> responseObserver) {
+    execute(
+        () -> {
+          TransactionState state = manager.getState(request.getTransactionId());
+          responseObserver.onNext(
+              GetTransactionStateResponse.newBuilder()
+                  .setState(ProtoUtil.toTransactionState(state))
+                  .build());
+          responseObserver.onCompleted();
+        },
+        responseObserver);
+  }
+
+  private DistributedTransaction getTransaction(String transactionId) throws TransactionException {
+    ActiveTransaction activeTransaction = activeTransactions.get(transactionId);
+    if (activeTransaction == null) {
+      throw new TransactionException(
+          "the transaction is not found. transactionId: " + transactionId);
+    }
+    return activeTransaction.get();
+  }
+
+  private static void execute(
+      ThrowableRunnable<Throwable> runnable, StreamObserver<?> responseObserver) {
+    try {
+      runnable.run();
+    } catch (IllegalArgumentException | IllegalStateException e) {
+      LOGGER.error("an invalid argument error happened during the execution", e);
+      responseObserver.onError(
+          Status.INVALID_ARGUMENT.withDescription(e.getMessage()).asRuntimeException());
+    } catch (CrudConflictException | CommitConflictException e) {
+      responseObserver.onError(
+          Status.FAILED_PRECONDITION.withDescription(e.getMessage()).asRuntimeException());
+    } catch (UnknownTransactionStatusException e) {
+      LOGGER.error("an unknown transaction error happened during the execution", e);
+      responseObserver.onError(Status.UNKNOWN.withDescription(e.getMessage()).asRuntimeException());
+    } catch (Throwable e) {
```

Comment:
I'm wondering Errors should be caught here.
I've never caught Errors but should it rather stop the server or something if an Error occurs?

Reply:
I catch `Throwable` here is because whenever any `Throwable` is thrown, I thought we should call `responseObserver.onError`, otherwise the client will get stuck. And for OutOfMemoryError, we have the `-XX:OnOutOfMemoryError` JVM flag, and we can kill the server with it. What do you think?